File: src/commands/log.rs
//! `gmeta log` — walk the commit history and print metadata for each commit,
//! matching the output of scripts/log.rb.

use anyhow::{Context, Result};
use git2::{Oid, Repository, Sort};

use crate::context::CommandContext;
use crate::db::Db;
use crate::git_utils;

// ── ANSI colours ──────────────────────────────────────────────────────────────
const RESET: &str = "\x1b[0m";
const BOLD: &str = "\x1b[1m";
const DIM: &str = "\x1b[2m";
const YELLOW: &str = "\x1b[33m";
const GREEN: &str = "\x1b[32m";
const CYAN: &str = "\x1b[36m";
const BLUE: &str = "\x1b[34m";

pub fn run(
    start_ref: Option<&str>, // commit-ish to start from (default HEAD)
    count: usize,            // max commits to show
    metadata_only: bool,     // skip commits with no metadata
) -> Result<()> {
    let ctx = CommandContext::open_git2(None)?;
    let repo = ctx.git2_repo()?;
    // log needs a separate Db with an embedded repo for blob resolution during reads
    let db_path = git_utils::git2_db_path(repo)?;
    let db = Db::open_with_repo(&db_path, git_utils::git2_discover_repo()?)?;

    // Resolve start ref → OID
    let start_oid = resolve_start(repo, start_ref)?;

    // Walk commits
    let mut revwalk = repo.revwalk()?;
    revwalk.set_sorting(Sort::TIME)?;
    revwalk.push(start_oid)?;

    let mut printed = 0usize;

    for oid_result in revwalk {
        if printed >= count {
            break;
        }

        let oid: Oid = oid_result.context("error walking commits")?;
        let commit = repo.find_commit(oid)?;
        let sha = oid.to_string();

        // Fetch metadata before deciding whether to print the commit
        let entries = db.get_all("commit", &sha, None).unwrap_or_default();
        // get_all returns (key, value, value_type, is_git_ref)
        // value is a JSON-encoded string for string types
        let meta: Vec<(String, String)> = entries
            .into_iter()
            .map(|(key, value, _vtype, _is_ref)| {
                // Decode JSON string wrapper → raw value for display
                let raw = serde_json::from_str::<String>(&value).unwrap_or(value);
                (key, raw)
            })
            .collect();

        if metadata_only && meta.is_empty() {
            continue;
        }

        // Blank line between entries
        if printed > 0 {
            println!();
        }
        printed += 1;

        // ── Header line ───────────────────────────────────────────────────────
        let short_sha = &sha[..10];
        let author = commit.author();
        let author_name = author.name().unwrap_or("unknown");
        let author_email = author.email().unwrap_or("");

        println!(
            "{YELLOW}commit {short_sha}{RESET} {DIM}—{RESET} \
             {GREEN}{author_name}{RESET} {DIM}<{author_email}>{RESET}"
        );

        // ── Commit message (first 4 non-empty lines) ──────────────────────────
        let message = commit.message().unwrap_or("").trim().to_string();
        let nonempty_lines: Vec<&str> = message.lines().filter(|l| !l.trim().is_empty()).collect();
        let shown = nonempty_lines.len().min(4);
        for line in &nonempty_lines[..shown] {
            println!("  {line}");
        }
        if nonempty_lines.len() > 4 {
            let extra = nonempty_lines.len() - 4;
            println!("  {DIM}... ({extra} more lines){RESET}");
        }

        // ── Metadata block ────────────────────────────────────────────────────
        if !meta.is_empty() {
            println!("  {CYAN}╶── metadata ──{RESET}");
            for (key, value) in &meta {
                let preview = format_value_preview(value);
                println!("  {BLUE}│{RESET} {BOLD}{key}{RESET}  {preview}");
            }
            println!("  {BLUE}╵{RESET}");
        }
    }

    if printed == 0 {
        if metadata_only {
            println!("No commits with metadata found.");
        } else {
            println!("No commits found.");
        }
    }

    Ok(())
}

// ── Helpers ───────────────────────────────────────────────────────────────────

/// Resolve a ref name or commit-ish to an OID.  Falls back to HEAD.
fn resolve_start(repo: &Repository, start_ref: Option<&str>) -> Result<Oid> {
    let spec = start_ref.unwrap_or("HEAD");
    repo.revparse_single(spec)
        .with_context(|| format!("could not resolve ref '{}'", spec))?
        .peel_to_commit()
        .with_context(|| format!("'{}' does not point to a commit", spec))?
        .id()
        .pipe(Ok)
}

/// Format a raw (already-decoded) metadata value for display.
/// Mirrors the Ruby format_value_preview function:
///   - JSON arrays   → "[list: N items]"
///   - JSON objects  → "{object: N keys}"
///   - strings       → first line, truncated to 50 chars; " ..." appended
///     if there are more lines
fn format_value_preview(value: &str) -> String {
    // Try JSON parse for arrays/objects
    if let Ok(json) = serde_json::from_str::<serde_json::Value>(value) {
        if let Some(arr) = json.as_array() {
            return format!("{CYAN}[list: {} items]{RESET}", arr.len());
        }
        if let Some(obj) = json.as_object() {
            return format!("{CYAN}{{object: {} keys}}{RESET}", obj.len());
        }
    }

    // Plain string
    let first_line = value.lines().next().unwrap_or("");
    let has_more_lines = value.contains('\n') && value.trim_end_matches('\n') != first_line;

    let mut preview = if first_line.len() > 50 {
        format!("{}...", &first_line[..50])
    } else {
        first_line.to_string()
    };

    if has_more_lines && first_line.len() <= 50 {
        preview.push_str(" ...");
    }

    format!("{DIM}{preview}{RESET}")
}

// Simple extension trait to pipe a value through Ok() without a temp binding
trait Pipe: Sized {
    fn pipe<T>(self, f: impl FnOnce(Self) -> T) -> T {
        f(self)
    }
}
impl<T> Pipe for T {}

File: scripts/log.rb
#!/usr/bin/env ruby

# Pretty-prints git log with git meta metadata for each commit.
#
# For each commit, shows: SHA, author, first 4 lines of the message,
# then all metadata keys with a preview of the value.
#
# Usage: ruby scripts/log.rb [--path=REPO_PATH] [-n COUNT]

require 'json'
require 'optparse'

repo_path = "."
count = 20
metadata_only = false

parser = OptionParser.new do |opts|
  opts.banner = "Usage: ruby scripts/log.rb [--path=REPO_PATH] [-n COUNT] [--mo]"
  opts.on("--path=PATH", "Path to the git repository (default: .)") { |p| repo_path = p }
  opts.on("-n COUNT", Integer, "Number of commits to show (default: 20)") { |n| count = n }
  opts.on("--mo", "Metadata only: show only commits that have metadata") { metadata_only = true }
end
parser.parse!(ARGV)

# ANSI color helpers
module C
  RESET   = "\e[0m"
  BOLD    = "\e[1m"
  DIM     = "\e[2m"
  YELLOW  = "\e[33m"
  GREEN   = "\e[32m"
  CYAN    = "\e[36m"
  MAGENTA = "\e[35m"
  BLUE    = "\e[34m"
  WHITE   = "\e[37m"
  RED     = "\e[31m"

  def self.yellow(s)  "#{YELLOW}#{s}#{RESET}" end
  def self.green(s)   "#{GREEN}#{s}#{RESET}" end
  def self.cyan(s)    "#{CYAN}#{s}#{RESET}" end
  def self.magenta(s) "#{MAGENTA}#{s}#{RESET}" end
  def self.blue(s)    "#{BLUE}#{s}#{RESET}" end
  def self.dim(s)     "#{DIM}#{s}#{RESET}" end
  def self.bold(s)    "#{BOLD}#{s}#{RESET}" end
  def self.red(s)     "#{RED}#{s}#{RESET}" end
end

def format_value_preview(value)
  # Try to parse as JSON first
  parsed = begin
    JSON.parse(value)
  rescue
    nil
  end

  if parsed.is_a?(Array)
    "#{C.magenta("[list: #{parsed.length} items]")}"
  elsif parsed.is_a?(Hash)
    "#{C.magenta("{object: #{parsed.keys.length} keys}")}"
  else
    # Treat as string — show first 50 chars or until first newline
    str = value.to_s
    first_line = str.split("\n", 2).first || ""
    preview = first_line.length > 50 ? first_line[0...50] + "..." : first_line
    preview += " ..." if str.include?("\n") && first_line.length <= 50
    C.dim(preview)
  end
end

# Get commit list
log_format = "%H%x00%an%x00%ae%x00%B%x00"
raw = `git -C #{repo_path} log --format="#{log_format}" -n #{count} 2>&1`

if !$?.success?
  $stderr.puts "Error running git log: #{raw}"
  exit 1
end

# Parse commits — split on the record separator
commits = raw.split("\0\n").reject(&:empty?).each_slice(1).map do |chunk|
  line = chunk.join
  parts = line.split("\0", 4)
  next if parts.length < 4
  {
    sha: parts[0].strip,
    author: parts[1].strip,
    email: parts[2].strip,
    message: parts[3].strip
  }
end.compact

if commits.empty?
  puts "No commits found."
  exit 0
end

printed = 0
commits.each do |commit|
  sha = commit[:sha]
  short_sha = sha[0...10]

  # Fetch metadata early so we can skip if --mo is set
  meta = nil
  meta_raw = `git meta get --json commit:#{sha} 2>/dev/null`
  if $?.success? && !meta_raw.strip.empty? && meta_raw.strip != "{}"
    meta = begin
      JSON.parse(meta_raw)
    rescue
      nil
    end
    meta = nil if meta && meta.empty?
  end

  next if metadata_only && meta.nil?

  puts "" if printed > 0
  printed += 1

  # Header
  puts "#{C.yellow("commit #{short_sha}")} #{C.dim("—")} #{C.green(commit[:author])} #{C.dim("<#{commit[:email]}>")}"

  # Message (first 4 lines)
  lines = commit[:message].split("\n").reject(&:empty?)
  msg_lines = lines[0...4]
  msg_lines.each do |line|
    puts "  #{line}"
  end
  if lines.length > 4
    puts "  #{C.dim("... (#{lines.length - 4} more lines)")}"
  end

  # Metadata display
  if meta
    puts "  #{C.cyan("╶── metadata ──")}"
    flatten_json("", meta).each do |key, value|
      preview = format_value_preview(value.is_a?(String) ? value : JSON.generate(value))
      puts "  #{C.blue("│")} #{C.bold(key)}  #{preview}"
    end
    puts "  #{C.blue("╵")}"
  end
end

# Recursively flatten nested JSON into dot-separated keys
BEGIN {
  def flatten_json(prefix, obj)
    results = []
    case obj
    when Hash
      obj.each do |k, v|
        full_key = prefix.empty? ? k : "#{prefix}:#{k}"
        if v.is_a?(Hash)
          results.concat(flatten_json(full_key, v))
        else
          results << [full_key, v]
        end
      end
    else
      results << [prefix, obj]
    end
    results
  end
}
